state-chain/pallets/cf-broadcast/src/lib.rs

```diff
@@ -1005,6 +1033,8 @@ impl<T: Config<I>, I: 'static> Broadcaster<T::TargetChain> for Pallet<T, I> {
 				}
 		}
 
+		IncomingKeyAndBroadcastId::<T, I>::set(Some((new_key, broadcast_id)));
```

Comment:
This approach to use `IncomingKeyAndBroadcastId` seems correct but I'm trying to think if there might be easier way to do it so we don't have to add yet another item to storage which requires another read to then delete it at `transaction_succeeded`.
I guess it could be possible to add a function for `ApiCall` to identify if the call is a key rotation and get the `new_key` from it if it is, right? The problem would then be that at `transaction_succeeded` we don't really have a way to know if it was a key rotation unless we store the `broadcastId`, which leads to the current approach. It just feels like simpler that we keep track of the key rotation through the system in the transaction itself rather than this.

The other solution I could come up with is that we implement logic in the engine to detect if transactions are key rotations and report that in `transaction_succeeded`. That would actually be fairly simple for Eth/Arb/Sol, which are the chains we need this logic for. But maybe it's just too cumbersome and the current approach is just simpler and therefore better?

Reply:
I dont think storage read/write is a problem because a rotation happens only once in 3 days and its just a few read/writes per rotation. I dont think I understand the second part of your comment.

Doing the recognition of rotation tx in the engine wont help us too much. The only thing we might, then, not need is the IncomingKeyAndBroadcastId. Other than that we still need all the functionality that enables us to store the signer inside the Apicall (to later compare the onchain key and the apicall signer). So not sure that little advantage is worth who knows how much effort in the engine.

Reply:
`IncomingKeyAndBroadcastId` is read in every `transaction_succeeded`  and the `CurrentOnChainKey` is read on every `start_next_broadcast_attempt`. If I'm not missing anything, that is for every transaction, not only in rotations.  My point was that those could be avoided if we did it in a clever way.
However, it's fair that a couple of extra reads for normal transactions and a couple more writes once in 3 days is fine to not have to overcomplicate it :+1:

Reply:
```IncomingKeyAndBroadcastId``` should not actually be read on every tx. The ```exists()``` function is supposed to be efficient in a way that it smartly/cheaply finds out if a storage item exists and only if it exists that we actually read the storage item. And because it would only exist around rotations, it would only be read for a few txs around that time. 

Its true that ```CurrentOnChainKey``` is read in ```start_next_broadcast_attempt``` but that is only called when a broadcast is retried and retry should only happen for a few txs. most of them should go through on first try. Also, at this point (```start_next_broadcast_attempt```) the information about what key is currently on the external chain has to be known and therefore has to be stored on the SC for the information to be available. I dont see how else would we retrieve this info (what the current onchain key is) on the fly. If we retrieve this info through some other storage item then that storage item has to be read so in any case we have to read storage.

Reply:
> The exists() function is supposed to be efficient in a way that it smartly/cheaply finds out if a storage item exists and only if it exists that we actually read the storage item
Alright, good to know!

Makes sense regarding the `CurrentOnChainKey` too, I was initially trying to not have to require the `IncomingKeyAndBroadcastId` but given what you said above it's not worth the effort. Therefore the current approach makes sense to me, thanks for providing extra info.